### DISCUSSION TITLE
Switch for push notifications

### DISCUSSION THREAD

**Discussion by mama31** [fwestenberg/reolink_dev#247] [Ideas]
Hello,

Is it possible to expose the push notifications to HA. Recording, FTP, email are working fine but I didn't find the push function.

Thanks a lot for all the great job already!

Hamad

**Comment by cpainchaud** [comment]
the Push API setting seems to be very buggy. I got it 'working' in theory but in practice, it just breaks the camera configuration. I contacted Reolink support to get a fix ...

**Comment by triggerx** [comment]
I see that the switch for Push Notifications seems to be "switchable", but does not seem to be reflected in the native Reolink app, and then at some point later, I see that it's been "re-enabled" in HA based on how it's set in the Reolink app.  Is this what you mean by being "very buggy?"  Hopefully it can be ironed out at some point.  Thanks!

**Comment by cpainchaud** [comment]
I fixed the issue after Reolink answered to my support request so it is working now.

That said the Push toggle exposed in API  is GENERAL and concerns all phones connected to the app. It is not accessible through the App, only API.

**Reply by triggerx** [reply]
Awesome, thanks for the quick response (and your work on the integration).  So, is the expected behavior that if I toggle the Notifications to OFF for a camera in HA using this integration, then no notifications will be sent, no matter what the app says on my mobile device, and that if I toggle the Notifications to ON in HA, then notifications will only be sent to mobile devices where the Notifications are enabled within the app?

**Comment by cpainchaud** [comment]
Take it as 2 in line switches

- API toggle which general and placed in front of all individual switches
- Phone app switch which is specific to your device and can be seen by your
device only

**Reply by triggerx** [reply]
Thanks again.. fully understand all that now.  I'll play around with it a bit now knowing this... but at first look, it seems as though the "push notifications" entity toggle switch isn't reflecting the push notifications schedule that I have running (set via local web address/gui of camera)... should it?

**Comment by mama31** [comment]
Hello.
It still KO on my side, the switch reflects the General push state but the toogle doesn't change the state (it comes back to the original state after one second).
I uninstalled - restarted - installed the reolink package to check if it was a cache issue.

**Comment by cpainchaud** [comment]
@mama31 then please provide useful informations and provide logs. thank you

**Reply by mama31** [reply]
Can you guide me on what level of logs would be needed. I tried to set debug, is it the proper one?

**Reply by mama31** [reply]
I tried something else :
I bypassed this part of the code in camera_api.py :
`        if 0 > 1 : print('toto') 
           #self._sw_version_object is not None and self._sw_version_object > ref_sw_version_3_0_0_0_0:
           #body = [{"cmd": "SetPushV20", "action": 0, "param": self._push_settings["value"]}]
           #body[0]["param"]["Push"]["enable"] = new_value
        else:
            body = [{"cmd": "SetPush", "action": 0, "param": self._push_settings["value"]}]
            body[0]["param"]["Push"]["schedule"]["enable"] = new_value`

And it works.
So I think it is the condition : `self._sw_version_object is not None and self._sw_version_object > ref_sw_version_3_0_0_0_0:` that was not working for me.

**Reply by cpainchaud** [reply]
what camera + software version are you using ?

enable logging:

```yaml
logger:
  default: warning
  logs:
    custom_components.reolink_dev: debug
    custom_components.reolink_dev.base.data: warning
    reolink: debug
```

**Reply by mama31** [reply]
I'm using two types of cameras

RLC-520
Config Version v3.0.0.0
Firmware Version v3.0.0.136_20121112

RLC-423
Config Version v3.0.0.0
Firmware Version v3.0.0.146_20122309

**Reply by cpainchaud** [reply]
can you please the output of your camera API:
curl -k  -v "https://x.x.x.x/cgi-bin/api.cgi?cmd=GetPush&user=admin&password=blabla"
and
curl -k  -v "https://x.x.x.x/cgi-bin/api.cgi?cmd=GetPushV20&user=admin&password=blabla"

thank you

**Reply by cpainchaud** [reply]
Try version 0.31 now, thank you

**Reply by mama31** [reply]
> can you please the output of your camera API: curl -k -v "https://x.x.x.x/cgi-bin/api.cgi?cmd=GetPush&user=admin&password=blabla" and curl -k -v "https://x.x.x.x/cgi-bin/api.cgi?cmd=GetPushV20&user=admin&password=blabla"
> 
> thank you

GetPush:
`[
  {
    "cmd": "GetPush",
    "code": 0,
    "value": {
      "Push": {
        "schedule": {
          "enable": 1,
          "table": "111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111"
        }
      }
    }
  }
]`

GetPushV20:
`[
  {
    "cmd": "Unknown",
    "code": 1,
    "error": {
      "detail": "not support",
      "rspCode": -9
    }
  }
]`

**Reply by cpainchaud** [reply]
Did you try version v0.31 ?

**Reply by mama31** [reply]
Hi! I just tried it!
Still KO, I looked into the reolink repository, I think you corrected the getPush but not the setPush.
So many thanks

**Reply by cpainchaud** [reply]
but is the GetPush working right? does it show the right status?

**Reply by cpainchaud** [reply]
Also I am still waiting for your detailed logs

**Reply by mama31** [reply]
The getpush was already working, it is the setPush that was the issue :
Here is a part of the log.
I can send you the full log privately but there is no way in Github I guess
`2021-10-22 13:12:32 DEBUG (MainThread) [reolink.camera_api] Sending command: SetPushV20 to: 192.168.1.50 with body: [{'cmd': 'SetPushV20', 'action': 0, 'param': {'Push': {'schedule': {'enable': 1, 'table': '111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111'}, 'enable': 0}}}]
2021-10-22 13:12:32 DEBUG (MainThread) [reolink.camera_api] send() HTTP Request params ={'cmd': 'SetPushV20', 'token': '469303462e583b2'}
2021-10-22 13:12:32 DEBUG (MainThread) [reolink.camera_api] send() HTTP Request body =[{'cmd': 'SetPushV20', 'action': 0, 'param': {'Push': {'schedule': {'enable': 1, 'table': '111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111'}, 'enable': 0}}}]
2021-10-22 13:12:32 DEBUG (MainThread) [reolink.camera_api] send() HTTP Response status=200 content-type=(text/html)
2021-10-22 13:12:32 DEBUG (MainThread) [reolink.camera_api.data] send() HTTP Response data: [
   {
      "cmd" : "Unknown",
      "code" : 1,
      "error" : {
         "detail" : "not support",
         "rspCode" : -9
      }
   }
]

2021-10-22 13:12:32 DEBUG (MainThread) [reolink.camera_api] Response from 192.168.1.50: [{'cmd': 'Unknown', 'code': 1, 'error': {'detail': 'not support', 'rspCode': -9}}]`

**Reply by mama31** [reply]
Hello,
I just managed to switch to the latest version and I still have the issue. I see the proper state of push but the toggle function doesn't work properly.

**Reply by mama31** [reply]
Hello. I've updated to v0.38 and I've lost completely the push switch.

**Comment by triggerx** [comment]
I've been able to confirm that switching between live camera view and the Settings page, in the Reolink App (iphone Xs, ios 15), will reset the push_notifications entity to "on" no matter what.  I can turn it off in HA, and it will stay off as long as I want.... but when I open the app, and it loads live view, then I touch the settings icon... within seconds, it sets push_notifications to On.  I can turn it of in HA, sit on the settings page forever and nothing changes, but when I go back to live view, and back to settings... .it turns push_notifications back on.

So, at this point, from what I can tell, it appears to be an issue in the Reolink app.  Happy to provide any additional information if needed.  Using a Reolink RLC-822A.

**Comment by cpainchaud** [comment]
I described the way it works on Android and also not the way Reolink described it which for once seemed to agree :) The toggle in API and phone app are independant, yet it may show 'off' in the app because the master switch is off but changing it in the App will have no effect on the API one.

**Reply by triggerx** [reply]
Agree, and completely understand how it should work. What I described what I'm seeing on the iPhone has nothing to do with what the toggle says in the app... the iPhone app is changing the master switch back to ON just by switching screens... regardless of what the toggle says in the app. So it seems like a bug Reolink needs to fix in their app, unfortunately.

**Comment by lightheaded** [comment]
Hi! This is a killer feature and I'd really like to automatically arm all my cameras when leaving home. 
> Take it as 2 in line switches - API toggle which general and placed in front of all individual switches - Phone app switch which is specific to your device and can be seen by your device only
Does this mean that I should have my app toggles always on, and just toggle Push via Home Assistant to disable/enable push? Will be able to test in the coming days, but just wanted to try to understand it.

**Reply by cpainchaud** [reply]
It's exactly what I am doing currently at my home.
